src/ParallelAlgorithms/Events/ObserveDataBox.hpp

```diff
@@ -0,0 +1,159 @@
+
+namespace Events {
+
+namespace detail {
+
+struct map_add {
+  std::map<std::string, size_t> operator()(
+      std::map<std::string, size_t> map_1,
+      const std::map<std::string, size_t>& map_2) {
+    for (const auto& [key, value] : map_2) {
+      map_1.at(key) += value;
+    }
+    return map_1;
+  }
+};
+
+using ReductionType = Parallel::ReductionData<
+    // Time
+    Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
+    // Map of total mem usage in MB per item in DataBoxes
+    Parallel::ReductionDatum<std::map<std::string, size_t>, map_add>>;
+
+template <typename ContributingComponent>
+struct ReduceDataBoxSize {
+  template <typename ParallelComponent, typename DbTags, typename Metavariables,
+            typename ArrayIndex>
+  static void apply(db::DataBox<DbTags>& /*box*/,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/, const double time,
+                    const std::map<std::string, size_t>& item_sizes) {
+    auto& observer_writer_proxy = Parallel::get_parallel_component<
+        observers::ObserverWriter<Metavariables>>(cache);
+    const std::string subfile_name =
+        "/ObserveDataBox/" + pretty_type::name<ContributingComponent>();
+    std::vector<std::string> legend;
+    legend.reserve(item_sizes.size() + 1);
+    legend.emplace_back("Time");
+    std::vector<double> columns;
+    columns.reserve(item_sizes.size() + 1);
+    columns.emplace_back(time);
+    const double scaling = 1.0 / 1048576.0;  // so size is in MB
+    for (const auto& [name, size] : item_sizes) {
+      legend.emplace_back(name);
+      columns.emplace_back(scaling * static_cast<double>(size));
+    }
+    Parallel::threaded_action<
+        observers::ThreadedActions::WriteReductionDataRow>(
+        // Node 0 is always the writer
+        observer_writer_proxy[0], subfile_name, legend,
+        std::make_tuple(columns));
+  }
+};
+
+struct ContributeDataBoxSize {
+  template <typename ParallelComponent, typename DbTags, typename Metavariables,
+            typename ArrayIndex>
+  static void apply(db::DataBox<DbTags>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& array_index, const double time) {
+    const auto& my_proxy =
+        Parallel::get_parallel_component<ParallelComponent>(cache)[array_index];
+    auto& target_proxy = Parallel::get_parallel_component<
+        observers::ObserverWriter<Metavariables>>(cache);
+    const auto item_sizes = box.size_of_items();
+    if constexpr (Parallel::is_singleton_v<ParallelComponent>) {
+      Parallel::simple_action<ReduceDataBoxSize<ParallelComponent>>(
+          target_proxy[0], time, item_sizes);
+    } else {
+      Parallel::contribute_to_reduction<ReduceDataBoxSize<ParallelComponent>>(
+          ReductionType{time, item_sizes}, my_proxy, target_proxy[0]);
+    }
+  }
+};
+}  // namespace detail
+
+class ObserveDataBox : public Event {
```

Comment:
Could you add some doxygen comment so this shows up?

How difficult would it be to write a simple test that just verifies this works with say 1 singleton, 1 nodegroup, and 1 array (with 3 elements?)? If it's difficult, I'm happy to defer since this feature is very necessary